index/internal/indexer/indexer.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2025 EngFlow Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexer
+
+import (
+	"encoding/json"
+	"log"
+	"maps"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"github.com/EngFlow/gazelle_cc/index/internal/collections"
+	"github.com/bazelbuild/bazel-gazelle/label"
+)
+
+type (
+	// Represents information about structure of possibly external dependency
+	Module struct {
+		// Name of external repository, or empty if targets are defined in the same Bazel repository
+		Repository string
+		// List of targets defined in given module, typically a single cc_library
+		Targets []*ModuleTarget
+	}
+	// Defines information about structure of rule that might be indexed, typically based on cc_library
+	ModuleTarget struct {
+		Name               label.Label
+		Hdrs               collections.Set[label.Label] // header files (each header is represented as a Label)
+		Includes           collections.Set[string]      // list of include paths
+		StripIncludePrefix string                       // optional prefix to remove
+		IncludePrefix      string                       // optional prefix to add
+		Deps               collections.Set[label.Label] // dependencies on other targets
+	}
+)
+
+type IndexingResult struct {
+	// Headers mapping to exactly one Bazel rule
+	HeaderToRule map[string]label.Label
+	// Headers defined in multiple rules
+	Ambiguous map[string]collections.Set[label.Label]
+}
+
+// Process list of modules to create an unfiorm index mapping header to exactly one rule that provides their definition.
+// In case if multiple modules define same headers might try to select one that behaves as clousers over remaining ambigious rules.
+func CreateHeaderIndex(modules []Module) IndexingResult {
+	// headersMapping will store header paths to a collections.Set of Labels.
+	headersMapping := map[string]*collections.Set[label.Label]{}
```

Comment:
Nit: remove the pointer to make the type `map[string]collections.Set[label.Label]`. `collections.Set` is already a reference type, so wrapping that in a pointer makes it a little confusing.

Or maybe `map[string][]label.Label` since it looks like there shouldn't be duplicate target labels.